### PR TITLE
presently possible rewrites for potential collect deprecation

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request rewrites (all?) `collect` calls that can presently be rewritten as `Vector`|`Matrix`|`Array`, constituting a head start on the potential `collect` deprecation between 0.7 and 1.0. (For the latest re. the potential `collect` deprecation, ref. https://github.com/JuliaLang/julia/issues/16029#issuecomment-355852101.) Best!

### CODE REVIEW

File: test/abstractarray.jl [hunk 1]
```diff
@@ -791,7 +791,7 @@ for A in (rand(2), rand(2,3))
     for (i, v) in pairs(A)
         @test A[i] == v
     end
-    @test collect(values(A)) == collect(A)
+    @test Array(values(A)) == Array(A)
```
Comment (nalimilan) [inline test/abstractarray.jl]: `Array` doesn't seem needed on the RHS?

File: test/arrayops.jl [hunk 2]
```diff
@@ -468,7 +468,7 @@ end
     @test find(c -> c == 'l', s) == [3]
     g = Base.Unicode.graphemes("日本語")
     @test find(isascii, g) == Int[]
-    @test find(!iszero, (i % 2 for i in 1:10)) == collect(1:2:9)
+    @test find(!iszero, (i % 2 for i in 1:10)) == Vector(1:2:9)
```
Comment (nalimilan) [inline test/arrayops.jl]: `Vector` shouldn't be needed now.

File: test/arrayops.jl [hunk 3]
```diff
@@ -1631,7 +1631,7 @@ end
     val, state = next(itr, state)
     @test done(itr, state)
     @test r[val] == 3
-    r = sparse(collect(2:3:8))
+    r = sparse(Vector(2:3:8))
```
Comment (nalimilan) [inline test/arrayops.jl]: `Vector` doesn't seem to be needed.

File: test/offsetarray.jl [hunk 4]
```diff
@@ -123,16 +123,16 @@ S = view(A, :, :)
 @test_throws BoundsError S[1,1]
 @test axes(S) === (0:1, 3:4)
 # https://github.com/JuliaArrays/OffsetArrays.jl/issues/27
-g = OffsetArray(collect(-2:3), (-3,))
+g = OffsetArray(Vector(-2:3), (-3,))
 gv = view(g, -1:2)
 @test axes(gv, 1) === Base.OneTo(4)
-@test collect(gv) == collect(-1:2)
+@test collect(gv) == Vector(-1:2)
```
Comment (nalimilan) [inline test/offsetarray.jl]: Not needed here not below.

File: test/random.jl [hunk 5]
```diff
@@ -483,9 +483,9 @@ let mta = MersenneTwister(42), mtb = MersenneTwister(42)
     @test randsubseq(mta,1:10,0.4) == randsubseq(mtb,1:10,0.4)
     @test randsubseq!(mta,Int[],1:10,0.4) == randsubseq!(mtb,Int[],1:10,0.4)
 
-    @test shuffle(mta,collect(1:10)) == shuffle(mtb,collect(1:10))
-    @test shuffle!(mta,collect(1:10)) == shuffle!(mtb,collect(1:10))
-    @test shuffle(mta,collect(2:11)) == shuffle(mtb,2:11)
+    @test shuffle(mta,Vector(1:10)) == shuffle(mtb,collect(1:10))
```
Comment (nalimilan) [inline test/random.jl]: Also change `collect` on the RHS.

File: test/sparse/sparse.jl [hunk 6]
```diff
@@ -1864,7 +1864,7 @@ end
 # are called. (Issue #18705.) EDIT: #19239 unified broadcast over a single sparse matrix,
 # eliminating the former operation classes.
 @testset "issue #18705" begin
-    S = sparse(Diagonal(collect(1.0:5.0)))
+    S = sparse(Diagonal(Vector(1.0:5.0)))
```
Comment (nalimilan) [inline test/sparse/sparse.jl]: `Vector` not needed, nor below.

File: test/statistics.jl [hunk 7]
```diff
@@ -338,9 +338,9 @@ end
     @test quantile([1,2,3,4],0.5) == 2.5
     @test quantile([1,2,3,4],[0.5]) == [2.5]
     @test quantile([1., 3],[.25,.5,.75])[2] == median([1., 3])
-    @test quantile(100.0:-1.0:0.0, 0.0:0.1:1.0) == collect(0.0:10.0:100.0)
-    @test quantile(0.0:100.0, 0.0:0.1:1.0, sorted=true) == collect(0.0:10.0:100.0)
-    @test quantile(100f0:-1f0:0.0, 0.0:0.1:1.0) == collect(0f0:10f0:100f0)
+    @test quantile(100.0:-1.0:0.0, 0.0:0.1:1.0) == Vector(0.0:10.0:100.0)
```
Comment (nalimilan) [inline test/statistics.jl]: `Vector` not needed.

**Review by nalimilan — COMMENTED** [review]
Thanks! Looks good, though I've noted a few places where `Vector` could be removed.

**Comment by Sacha0** [comment]
Thanks for the thorough review @nalimilan! Comments addressed :).

**Comment by timholy** [comment]
I'm still a little bit worried about https://github.com/JuliaLang/julia/issues/16029#issuecomment-342648509. Is the idea that we've made a stronger distinction between `convert(T, x)` and `T(x)`? (Sorry, I haven't had time to follow all the issues.) `convert(Vector, x)` should *definitely* fail if `x` doesn't start indexing with 1, but if we're being clearer about the distinction then I can see `Vector(x)` being justified.

**Review by nalimilan — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
@JeffBezanson: can you clarify the new relationship between `T(x)` and `convert(T, x)`?

**Comment by JeffBezanson** [comment]
The differences between `convert` and constructors are:

1. The one we've always had: `convert` should only be for "sufficiently similar" things.
2. `convert(::Type{T}, x::T)` can and should just return `x`, it doesn't need to copy.

So it is still the case that constructors have more leeway than `convert`. However, we would need to be careful that `convert` in those cases does extra checks before calling the constructor.

**Comment by Sacha0** [comment]
(Marking triage for a decision re. https://github.com/JuliaLang/julia/issues/16029#issuecomment-355852101 / this pull request. Best!)

**Comment by mschauer** [comment]
I think these changes are meaningful even if `collect` is kept and do not need to wait for a decision.

**Comment by Sacha0** [comment]
> I think these changes are meaningful even if collect is kept and do not need to wait for a decision.

After reviewing these changes again, I tend to agree :). So absent objections or requests for time, I plan to merge these changes tomorrow midday PT or later. Best!

**Comment by Sacha0** [comment]
(AV i686 timed out without apparent issue prior, Travis i686 hit an unrelated libgit2 failure without other apparent issues, and Travis macOS hit an unrelated spawn failure without other apparent issues.)

**Comment by Sacha0** [comment]
Thanks all! :)

**Pull request merged**